File: api/src/main/java/com/void01/bukkit/voidframework/api/common/library/Dependency.java
package com.void01.bukkit.voidframework.api.common.library;

import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Getter
public class Dependency {
    @NonNull
    private final String groupId;
    @NonNull
    private final String artifactId;
    @NonNull
    private final String version;

    public String getAsGradleStyleExpression() {
        return groupId + ":" + artifactId + ":" + version;
    }

    public static Dependency fromGradleStyleExpression(@NonNull String expression) {
        String[] split = expression.split(":");

        if (split.length != 3) {
            throw new IllegalArgumentException("Invalid expression");
        }

        return new Dependency(split[0], split[1], split[2]);
    }
}

File: dependency/Dependency.kt
package com.github.julyss2019.bukkit.voidframework.dependency

class Dependency(val groupId: String, val artifactId: String, val version: String) {
    companion object {
        fun create(groupId: String, artifactId: String, version: String): Dependency {
            return Dependency(groupId, artifactId, version)
        }

        @JvmStatic
        fun createByGradleCoordinates(coordinates: String): Dependency {
            val split = coordinates.split(":")

            require(split.size == 3) { "invalid coordinates" }

            return Dependency(split[0], split[1], split[2])
        }
    }

    override fun toString(): String {
        return "Dependency(groupId='$groupId', artifactId='$artifactId', version='$version')"
    }
}
